Architecting templates (#167)
aws-samples/cfn101-workshop: content/intermediate/templates/architecting-templates/index.md

---
title: "Architecting your templates"
weight: 650
---

> ### Overview

Reviewer: Test

Create a new stack with the `base-network.template` file:

1. Navigate to the [AWS CloudFormation Console](https://console.aws.amazon.com/cloudformation/).
2. From the region selector on the top of the page, choose a region - such as *N. Virginia* (`us-east-1`).
3. From **Create stack**, choose **With new resources (standard)**.
4. From **Prepare template**, choose **Template is ready**.
5. From **Template source**, choose **Upload a template file**. Choose the `base-network.template` file mentioned earlier, and then choose **Next**.
6. Specify a stack name: `cloudformation-workshop-dev-base-network`. Choose to accept default parameter values, and then choose **Next**.
7. On the subsequent page, choose **Next**.
8. Choose **Create stack** in the next page.

Stack creation will start: at the end of it, you should see your stack in the `CREATE_COMPLETE` status. As the creation of your stack proceeds, open the `base-network.template` file in your favorite text editor in your workstation. Note the following:

> * As you want to create your subnets in separate availability zones within a region (for example, `us-east-1a`, `us-east-1b` availability zones in the `us-east-1` region), you specify the `AvailabilityZone` [property](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-subnet.html) for each subnet. Instead of hard-coding the availabilty zone name (such as `us-east-1a`), the example template uses the `Fn::Sub` [intrinsic function](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-sub.html) (shown in the template using its YAML short form, `!Sub`) to concatenate the name of the region (e.g., `us-east-1`, derived from the `AWS::Region` [pseudo parameter](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/pseudo-parameter-reference.html#cfn-pseudo-param-region)) with either `a` or `b`, such as with: `AvailabilityZone: !Sub '${AWS::Region}a'`. This way, it is easier for you to reuse this template across regions by making the template more portable.

Reviewer 2: ```suggestion
* As you want to create your subnets in separate availability zones within a region (for example, `us-east-1a`, `us-east-1b` availability zones in the `us-east-1` region), you specify the `AvailabilityZone` [property](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-subnet.html) for each subnet. Instead of hard-coding the availability zone name (such as `us-east-1a`), the example template uses the `Fn::Sub` [intrinsic function](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-sub.html) (shown in the template using its YAML short form, `!Sub`) to concatenate the name of the region (e.g., `us-east-1`, derived from the `AWS::Region` [pseudo parameter](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/pseudo-parameter-reference.html#cfn-pseudo-param-region)) with either `a` or `b`, such as with: `AvailabilityZone: !Sub '${AWS::Region}a'`. This way, it is easier for you to reuse this template across regions by making the template more portable.
```

Reviewer 2: Not sure why the changed word is not highlighted..it just a typo in `availabilty` -> `availability`

Congratulations! You have successfully performed and verified the deployment of a sample application and of its infrastructure with CloudFormation!

### Challenge

In this challenge, you'll recall and apply key concepts related to reuse and modularity, and extend them in terms of deployment orchestration. You are tasked with copying the `application.template` file (that is in your workspace in Cloud9) into a new file called `application-blue-green.template`, and to update this new file so that you can use it for 2 stacks in the context of a blue/green deployment pattern where you'll [Update DNS Routing with Amazon Route 53](https://docs.aws.amazon.com/whitepapers/latest/blue-green-deployments/update-dns-routing-with-amazon-route-53.html). Requirements in this example challenge are as follows:

* Have 2 stacks: update the existing one you created, called `cloudformation-workshop-dev-application`, and create a new one called `cloudformation-workshop-dev-application-v2`, showing `Blue` and `Green` as outputs, respectively, instead of `Hello world!`. Let's refer to both stacks as to *Blue* and *Green*, respectively. Use the new `application-blue-green.template` file for both stacks.

> * Each stack should create a new alias record whose `Name` is the same to the one you used earlier (the hosted zone name). Both of these records though will need to be a [weighted set](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-values-weighted.html#rrsets-values-weighted-weight), so that Route 53 will respond to queries you issue based on the [ratio of the weight for a given resource to the total of weights](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy-weighted.html). Initially, you want to assign a weight of `255` (more weight) to the *Blue* stack, and `1` (less weight) to the *Green* stack: this way, *Blue* initially gets 255/256ths of traffic, and *Green* gets 1/256ths of traffic.

Reviewer 2: ```suggestion
* Each stack should create a new alias record whose `Name` is the same to the one you used earlier (the hosted zone name). Both of these records though will need to be a [weighted set](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-values-weighted.html#rrsets-values-weighted-weight), so that Route 53 will respond to query you issue based on the [ratio of the weight for a given resource to the total of weights](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy-weighted.html). Initially, you want to assign a weight of `255` (more weight) to the *Blue* stack, and `1` (less weight) to the *Green* stack: this way, *Blue* initially gets 255/256ths of traffic, and *Green* gets 1/256ths of traffic.
```

Author: Whilst I'd rather keep `queries` (plural) here - as the user will perform more than one query - I will accept this suggestion of using `query` and, in a subsequent commit, I will prepend `a` to it.

:::code{language=shell showLineNumbers=false showCopyAction=true}
aws cloudformation update-stack \
    --stack-name cloudformation-workshop-dev-application \
    --template-body file://application-blue-green.template \
    --parameters \
        ParameterKey=PageTextContent,ParameterValue=Blue \
        ParameterKey=RecordSetWeight,ParameterValue=1

aws cloudformation update-stack \
    --stack-name cloudformation-workshop-dev-application-v2 \
    --template-body file://application-blue-green.template \
    --parameters \
        ParameterKey=PageTextContent,ParameterValue=Green \
        ParameterKey=RecordSetWeight,ParameterValue=255
:::

* Once stack updates are complete, test that the `my-example-domain.com` record now mainly points to `Green`:

:::code{language=shell showLineNumbers=false showCopyAction=true}
curl http://my-example-domain.com
:::

::::

### Clean up

Navigate to the CloudFormation Console on your workstation. Select and delete, in the following order, stacks you created in this lab. *Note that since you used cross-stack references a number of stacks, you cannot delete a stack that exports a value if that value is still used by a consuming stack*:

> * Delete `cloudformation-workshop-dev-application-v2` and `cloudformation-workshop-dev-application` without waiting for each one of them has been deleted: such stacks are not depending between each others. When both stacks are deleted, continue with next steps.
> * Delete `cloudformation-workshop-dev-security-groups` and `cloudformation-workshop-dev-hosted-zone` without waiting for each one of them has been deleted as well. Then, continue with next steps.
> * `cloudformation-workshop-dev-cloud9`: this should delete also another stack whose name starts with `aws-cloud9-aws-cloudformation-workshop-`. When both stacks are deleted, continue with the last step shown next.
> * `cloudformation-workshop-dev-base-network`.
> 

Reviewer 2: ```suggestion
1. Delete `cloudformation-workshop-dev-application-v2` and `cloudformation-workshop-dev-application` without waiting for each one of them to be deleted; such stacks do not depend on each other. When both stacks are deleted, continue with next step.
2. Delete `cloudformation-workshop-dev-security-groups` and `cloudformation-workshop-dev-hosted-zone` without waiting for each one of them to be deleted as well. Then, continue with next steps.
3. Delete `cloudformation-workshop-dev-cloud9`: this will delete Cloud9 instance. Also delete stack whose name starts with `aws-cloud9-aws-cloudformation-workshop-`. When both stacks are deleted, continue with the last step.
4. Delete `cloudformation-workshop-dev-base-network` stack.
```

Reviewer 2: As we refer to steps, I suggest changing bullet points to numbered list (I was about to use 1.,1....1, but my IDE auto-corrects that to 1,2,3,4) Feel free to change to 1,1,1,1 :)
